service/create/asg_stack.go

```diff
@@ -246,16 +257,55 @@ func (s *Service) updateASGStack(input asgStackInput) error {
 		Client: input.clients.CloudFormation,
 
 		// Settings.
-		ASGMaxSize:  input.asgSize,
-		ASGMinSize:  input.asgSize,
-		ImageID:     imageID,
-		Name:        key.AutoScalingGroupName(input.cluster, input.asgType),
-		TemplateURL: templateURL,
+		ASGMaxSize:            input.asgSize,
+		ASGMinSize:            input.asgSize,
+		ImageID:               imageID,
+		MaxBatchSize:          getMaxBatchSize(input.asgSize),
+		MinInstancesInService: getMinInstancesInService(input.asgSize),
+		Name: key.AutoScalingGroupName(input.cluster, input.asgType),
+		RollingUpdatePauseTime: rollingUpdatePauseTime,
+		TemplateURL:            templateURL,
+	}
+
+	stackUpdated, err := stack.Update()
+	if err != nil {
+		return stackUpdated, microerror.Mask(err)
+	}
+
+	return stackUpdated, nil
+}
+
+// getMaxBatchSize calculates the max batch size for the rolling update policy.
+func getMaxBatchSize(asgSize int) int {
+	var batchSize int
+
+	switch {
+	case asgSize <= 2:
+		batchSize = 1
+	case asgSize <= 4:
+		batchSize = 2
+	default:
+		// Calculate batch size and round up to nearest int.
+		result := float64(asgSize) * asgMaxBatchSizeRatio
+		batchSize = int(result + 0.5)
 	}
 
-	if err := stack.Update(); err != nil {
-		return microerror.Mask(err)
+	return batchSize
+}
+
+// getMinInstancesInService calculates the min number for instances to keep in
+// service for the rolling update policy.
+func getMinInstancesInService(asgSize int) int {
+	var minInstances int
+
+	switch {
+	case asgSize <= 2:
+		minInstances = 1
+	default:
+		// Calculate min instances and round up to nearest int.
+		result := float64(asgSize) * asgMinInstancesRatio
+		minInstances = int(result + 0.5)
```

Comment:
Yes, the rolling update is very different for AWS compared to KVM. From my POV using an autoscaling group lets us manage the nodes as a pool of resources. If we don't use an ASG then we need to implement extra logic like checking for and replacing failed nodes. Other projects like kops and our aws-terraform also use ASGs.

But it's a tradeoff and some features of the cluster TPO like node IDs are hard to implement.The tags on the EC2 nodes come from the ASG so they have the cluster ID but not the node IDs. I think we could tag the EC2 nodes ourselves but we'd need to do that in the operator.

For the kubernetesd update I was thinking that it would update the CoreOS AMIs for all nodes. The operator will then implement that change via the rolling upgrade.

I like the idea of a controller scheduling the batch operations. Especially as these are long running events due to the 5 min pause time that's needed between each batch to allow new nodes to join. This could also handle adding the extra nodes for the rolling upgrade and removing them afterwards.